url026791d51b_dhleong_SeriesTracker_v6_tgz-pJ8-stv6_SeriesTrackerJ8/src/stv6/database/MysqlDatabase.java

```diff
@@ -0,0 +1,994 @@
+		try {
+			PreparedStatement stmt = prepare(QueryType.TRACK_UPDATE);
+			
+			for (TrackData d : data) {
+				stmt.setInt(1, d.seriesId);
+				stmt.setInt(2, d.userId);
+				stmt.setInt(3, d.episode);
+				stmt.setLong(4, d.lastView);
+				stmt.addBatch();
+			}
+			
+			// gogogo
+			wrapBatch(stmt);
+		} catch (SQLException e) {
+			e.printStackTrace();
+		} 
+	}
+
+	/**
+	 * @param idsTable Either "series" or "users"
+	 */
+	private synchronized void updateIds(QueryType idsTable, QueryType trackQry, 
+			List<IdUpdateData> ids) {		
+		try {		
+			PreparedStatement idsStmt = prepare(idsTable);
+			PreparedStatement tracksStmt = prepare(trackQry);
+			
+			// work backwards; this should gracefully prevent 
+			//  duplicate id issues, since new ids will always
+			//  be increasing...
+			int oldId, newId;			
+			for (int i=ids.size()-1; i>=0; i--) {			
+				newId = ids.get(i).newId;
+				oldId = ids.get(i).oldId;
+				
+				idsStmt.setInt(1, newId);
+				idsStmt.setInt(2, oldId);
+				idsStmt.addBatch();				
+				//System.out.println(oldId + " => " + newId );
+				
+				tracksStmt.setInt(1, newId);
+				tracksStmt.setInt(2, oldId);
+				tracksStmt.addBatch();
+				//trackswrapUpdate(stmt);//.executeUpdate();
+			}
+			
+			wrapBatch(idsStmt, tracksStmt);
+		} catch (SQLException e) {
+			e.printStackTrace();
+		} 
+		
+	}
+
+	@Override
+	public synchronized void updateSeriesIds(List<IdUpdateData> ids) {
+		/*
+		updateIds(QueryType.SYNC_UPDATE_SERIES_IDS, QueryType.SYNC_UPDATE_TRACK_BY_SERIES, 
+				oldIds, newIds);
+		*/
+		
+		int startId = ids.get(ids.size()-1).oldId;
+		if (lastSeriesId > startId)
+			startId = createNewSeriesId();
+		
+		// first, direct oldIds to some unique spot; then, once there,
+		//  we put them into their final destination
+		List<IdUpdateData> oldToUnique = new ArrayList<IdUpdateData>();
+		List<IdUpdateData> uniqueToNew = new ArrayList<IdUpdateData>();
+		for (int i=0; i<ids.size(); i++, startId++) {
+			oldToUnique.add(new IdUpdateData(ids.get(i).oldId, startId));
+			uniqueToNew.add(new IdUpdateData(startId, ids.get(i).newId));
+		}
+		
+		// holy crap update
+		updateIds(QueryType.SYNC_UPDATE_SERIES_IDS, QueryType.SYNC_UPDATE_TRACK_BY_SERIES, oldToUnique);
+		updateIds(QueryType.SYNC_UPDATE_SERIES_IDS, QueryType.SYNC_UPDATE_TRACK_BY_SERIES, uniqueToNew);
+	}
+
+	/**
+	 * Set the last sync time for the given sync page on the given
+	 * 	SyncSettings to "now"
+	 * 
+	 * @param syncPage
+	 */
+	@Override
+	public synchronized void updateSyncTime(SyncSettings settings, SyncPage syncPage) {
+		long now = Profile.getNowSeconds();
+		
+		try {
+			PreparedStatement stmt = prepare(syncTimeQueries[syncPage.ordinal()][SYNC_TIME_UPDATE]);
+			stmt.setLong(1, now);
+			stmt.setString(2, settings.getBaseUrl());			
+			int updates = wrapUpdate(stmt);//.executeUpdate();
+			stmt.close();
+
+			if (updates == 0) {
+				// no existing entry; create it
+				stmt = prepare(syncTimeQueries[syncPage.ordinal()][SYNC_TIME_INIT]);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *stmt*. The resource is referenced at lines 917, 918, 919. The resource is closed at line 920. There are other execution paths that do not contain closure statements, for example, when MysqlDatabase.wrapUpdate throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).